Add CONTRIBUTING.md (#3085)
xournalpp/xournalpp: CONTRIBUTING.md

   period.

### Creating a fork

To make a code change, you'll need to first fork the repository on GitHub, if
you haven't already. To do so, click the "Fork" button in the top right corner
of the repository's web page.

We assume that you are already familiar with Git, which we use to track changes.
An excellent resource for Git is the _Pro Git_ book by Scott Chacon and Ben
Straub, which is available online for free
[here](https://git-scm.com/book/en/v2).

You should try to first clone your fork and then compile it manually. The
instructions for compiling Xournal++ on your operating system can be found at
[`LinuxBuild.md`](readme/LinuxBuild.md), [`MacBuild.md`](readme/MacBuild.md),
and [`WindowsBuild.md`](readme/WindowsBuild.md). Once you are set up, you will
be ready to make code changes.

### Code conventions and guidelines

Here are some of the _guidelines_ that we follow when working on the Xournal++
codebase. We emphasize _guidelines_ because these are ideal points that we'd
like to see in contributions, but we understand that perfectly following all
points is very difficult in practice.

General contribution guidelines:

> * **The code should be developed on a Linux distribution (of your choosing)**.
>   The project is set up to _build_ on MacOS or Windows but is not ideal for
>   development there.

Reviewer: Is this 'by design' or just because so far nobody cared or had the time to work out a decent development setup – or is there an inherent reason I am missing?

In the second case something can be done about this and I'd explicitly state something like

```suggestion
* **So far Linux is the most convenient platform for code development.** The
  documented setup on MacOS and Windows is not ideal, but if you have an idea,
  you're welcome to share it in a PR!
```

In the first case ... well, I'll have to dump my work on Windows console and crash handlers then (yes, this comment is completely self-serving (: ).

But jokes aside, my (admittedly unqualified) opinion is that, for a multi platform app, it helps if development happens on all target platforms.

Author: Thanks for the suggestion. The answer to your question is simply:

>  so far nobody cared or had the time to work out a decent development setup

The Windows setup relies on Msys2, which assumes familiarity with Linux. But the majority of people familiar with Linux are likely using Xournal++ on Linux rather than on Windows with Msys2.

> for a multi platform app, it helps if development happens on all target platforms.

I agree with you, but all the core maintainers are each using some Linux distribution as a daily driver. Due to limitations on time, the extent to which we "develop" on Windows and MacOS is limited to fixing platform-specific bugs and testing the packaging.

I think setting up development for Windows is actually not too bad: steps are just install dependencies & run build scripts to the developer. IDE configuration is undocumented, though.

On the other hand, trying to develop for MacOS means having to go through the complexity of setting up jhbuild and compiling GTK from source.

Reviewer: Thanks for your answer!

> Due to limitations on time, the extent to which we "develop" on Windows and MacOS is limited ...

This is understandable of course, but makes community input all the more important, doesn't it? That's why I suggest to use a more encouraging wording. I just think "You should use Linux" might put off potential contributors, which would be a pity because, as you said:

> setting up development for Windows is actually not too bad

... except for horrible build times with GCC, but this is fixable, see below.

> The Windows setup relies on Msys2, which assumes familiarity with Linux. But the majority of people familiar with Linux are likely using Xournal++ on Linux rather than on Windows with Msys2.

I wouldn't be so sure here. I believe I'm not the only one using Xournal++ on Windows as a simpler/more transparent/just better fitting alternative to OneNote for taking notes on lectures primarily; thus I'd rather not fiddle around with a Linux install on my Surface Book 2 just for development. Installing MSYS2 etc. is much simpler compared to that.
(That's not to say Linux is bad, I have a dual boot machine set up myself, just not the machine I use Xournal++ on. And coding on one machine, then packaging and transferring the whole build to the other to test it will get tedious.)

> IDE configuration is undocumented, though.

I can try to fix this once I've got some more time and figured out the last details.

<details><summary>(OT) Sneak peek: developing on Windows – will submit/update PRs when ready...</summary>
I'm using VS Code with C++, clangd, CMake extensions and build with Clang from the MSYS2/clang64 environment, which builds much faster than GCC, but needed some adjustments with the packaging script. Three compiler/linker flags later Clang emits PDB symbols, so now Xournal++ can be debugged with the apparently so popular VS debugger (compared to gdb it's really ... fast).<br>
Along the way this enabled rich stacktraces using the Win32 dbghelp API, so with crash handlers  implemented (which I did for #3746 initially), Windows users will also be able to file meaningful bug reports (#3802 being a counterexaple of 'meaningful').<br>
<b>TL;DR:</b> development on Windows does have some potential.</details>

Author: Thanks for the feedback, I incorporated it into the updated bullet point. Let's move the discussion into a separate issue, as I think it would be more visible to other people developing on Windows.

# Contributing to Xournal++

Thanks for your interest in contributing to Xournal++! The project is run
completely by volunteers in their spare time, so any contribution--no matter how
small--is greatly appreciated. In this file, we'll outline how you can
contribute to the project.

Xournal++ development primarily occurs on [GitHub at the xournalpp/xournalpp

> repository](https://github.com/xournalpp/xournalpp/issues/3065). As a

Reviewer 2: ```suggestion
repository](https://github.com/xournalpp/xournalpp). As a
```

fix, as well as a short description of what your changes are and why are you
making them. Feel free to also include questions or comments about code snippets
you may be unsure about.

### The review process

To maintain the quality of contributions, we typically require a PR to meet the
following requirements before it is merged.

* **The PR is reviewed by at least one maintainer**, and the relevant feedback is
  addressed.
* The code **successfully compiles** and passes all tests on our automated testing
  infrastructure.
* The code **passes the automated formatting check**. If the code does not pass,
  you can copy and paste the required changes from the Azure Pipelines page that
  has details on how the check failed.
* Before the PR is merged, a grace period of at least 24 hours have passed since
the last maintainer has approved the PR. This allows time for objections or
last-minute feedback. Larger or more complicated PRs may take a longer grace
period.

Once a PR is ready to merge, we typically ask submitters to rebase or squash
their commits in order to keep the change history cleaner. If you do not know

> how to do this, feel free to either ask for help.

Reviewer 2: This last sentence doesn't look complete to me. Is there an alternative to asking for help?

Author: Initially I wanted to say you could either 1) ask how to do it or 2) ask the maintainer to do it for you, but I changed it to just "asking for help" to simplicity. Just forgot to remove "either."

# Contributing to Xournal++

Thanks for your interest in contributing to Xournal++! The project is run
completely by volunteers in their spare time, so any contribution--no matter how
small--is greatly appreciated. In this file, we'll outline how you can
contribute to the project.

Xournal++ development primarily occurs on [GitHub at the xournalpp/xournalpp
repository](https://github.com/xournalpp/xournalpp). As a contributor, you

> probably have a particular bug or feature that are you interested in working on.

Reviewer 3: ```suggestion
probably have a particular bug or feature that you are interested in working on.
```

> Before you do anything, you should first look in the [issue

Reviewer 3: Sounds less harsh.
```suggestion
Before you start, you should first look in the [issue
```

tracker](https://github.com/xournalpp/xournalpp/issues) to see if anyone has
reported your bug or has a similar idea for a feature. If not, you should first
submit a new issue detailing what you are about to do. This will allow you to
get feedback and ensure that you do not end up duplicating work. You can obtain
additional help by contacting community members through one of our [official
communication channels](https://xournalpp.github.io/community/help/).

## Contributing translations

If you would like to contribute translations, you can submit improvements to

> [our project on Crowdin](https://crowdin.com/project/xournalpp).

Reviewer 3: Maybe add a few lines on when we merge these translations.

Reviewer 4: @LittleHuba like  that?
```suggestion
[our project on Crowdin](https://crowdin.com/project/xournalpp).
Those are then merged directly before the next release. 
So they wont appear directly after your contribution.
```

## Contributing code changes with pull requests (PRs)

To make a code change, you'll need to first fork the repository on GitHub, if
you haven't already. To do so, click the "Fork" button at the top right corner
of the repository's web page.

We assume that you are already familiar with Git, which we use to track changes.
An excellent resource for Git is the _Pro Git_ book by Scott Chacon and Ben
Straub, which is available online for free
[here](https://git-scm.com/book/en/v2).

You should try to first clone your fork and then compile it manually. The
instructions for compiling Xournal++ on your operating system can be found at
[`LinuxBuild.md`](readme/LinuxBuild.md), [`MacBuild.md`](readme/MacBuild.md),
and [`WindowsBuild.md`](readme/WindowsBuild.md). Once you are set up, you will
be ready to make code changes.

### Code conventions and guidelines

Here is a quick list of guidelines that we follow when working on the Xournal++
codebase.

> * Xournal++ is a mixture C++ and C code. This is unavoidable because we rely on

Reviewer 2: ```suggestion
* Xournal++ is a mixture of C++ and C code. This is unavoidable because we rely on
```

  many C libraries (e.g. GTK). Prefer C++ when possible, however.
* Although we do not have a strong opinion on code style, we use `clang-format`
  to enforce a _consistent_ code style. Feel free to write your code in whatever
  style you prefer, as long as you run `clang-format` to format your code

>   afterwards.

Reviewer 3: It would be great if you would provide our reasons for this (merge conflicts, hard to read code)

* Please accompany your code changes with documentation.
* Follow modern C++ best practices as listed in the [C++ Core
  Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines). In
  particular, make variables `const` by default, always use smart pointers

>   instead of new/delete, and prefer `std::optional` over null pointers.

Reviewer 4: ```suggestion
  instead of new/delete, and prefer `std::optional` over null pointers.
  A usefull tool to help you to fullfill those rules is clang-tidy. It can be included in nearly every modern IDE.
```

Reviewer 3: More in the lines of "they are merged from time to time"

* When working with code that interfaces with C libraries (GLib and GTK in
  particular), you should be extra careful with reference counting and memory

>   management.

Reviewer 4: ```suggestion
  management.
  It might be helpfull, to create own memory managing RAII structures. This will also reduce the time others must spend to verify your code, because it is correct by design.
```

* TODO, etc.

> 

Reviewer 4: Maybe add a hint, that it is appreciated to use clang-tidy.

Reviewer 2: Hmm, I never used clang-tidy. How to do that?

Reviewer 4: The easiest way to do so is to use clang-tidy over the IDE you are using. It is also possible, to enable it for every source file over the cmake variable `xournalpp_LINT`.

Reviewer 2: Got it. Just needed to install an extension in VS code. Thanks for the hint!

> 

Reviewer 3: Definitely add a bullet point for well-named variables. There is nothing worse than not knowing what that variable `x` means again.

Reviewer 4: ```
* Please use meaningful variable names. 
However, when the Type is already describing the variables purpose and if the context of the variable is very short it is fine to use short names. Also common abbreviations are welcome to be used. 
```

> 

Reviewer 3: We should also add a bullet point concerning the usage of `auto`.

Reviewer 4: ```
* Prefere to use `auto` over `Type` everywhere, where the type is derivable from the context.
* Especially if the type is long, like `std::vector<std::pari<std::string, MyLooooooongType>>`
* Prefer `auto&&` over `auto` and `auto const&` over `auto const`
```

Reviewer 3: Thas sounds good except the typo in std::pair

Reviewer 3: Without typos:

```
* Prefer to use `auto` over `Type` everywhere, where the type is derivable from the context.
* Especially if the type is long, like `std::vector<std::pair<std::string, MyLooooooongType>>`
* Prefer `auto&&` over `auto&` and `auto const&` over `auto const`
```

### Sending your contributions for review

> When you feel like your are ready to submit your code changes to be integrated

Reviewer 5: Shouldn't this be `you` instead of `your`?

> into the main Xournal++ repository, you will want to submit a pull request (PR).

Reviewer 3: ```suggestion
into the main Xournal++ repository or require some feedback on your work, you will want to submit a pull request (PR).
```

To do so, push your changes on to a branch of your fork in GitHub, and then hit
the "New pull request" button on the [main repository's PR
tracker](https://github.com/xournalpp/xournalpp/pulls). Select the branch you
want to contribute, and then press "Create pull request." If you feel like your
work isn't quite ready, feel free to submit your PR as a draft.

In your description, please include any relevant issue numbers that your PR will

> fix, as well as a short description of what your changes are and why are you

Reviewer 3: ```suggestion
fix, as well as a short description of what your changes are and why you are
```

making them. Feel free to also include questions or comments about code snippets
you may be unsure about.

### The review process

To maintain the quality of contributions, we typically require a PR to meet the
following requirements before it is merged.

* **The PR is reviewed by at least one maintainer**, and the relevant feedback is
  addressed.
* The code **successfully compiles** and passes all tests on our automated testing
  infrastructure.
* The code **passes the automated formatting check**. If the code does not pass,
  you can copy and paste the required changes from the Azure Pipelines page that
  has details on how the check failed.

> * Before the PR is merged, a grace period of at least 24 hours have passed since

Reviewer 3: ```suggestion
* Before the PR is merged, a grace period of at least 24 hours has passed since
```

small--is greatly appreciated. In this file, we'll outline how you can
contribute to the project.

## Overview

Several of the main ways that you can contribute to the project include:

* Updating [the website](https://github.com/xournalpp/xournalpp.github.io)
  with additional information.

> * Submitting translation improvements via crowdin (see below).

Reviewer 2: ```suggestion
* Submitting translation improvements via Crowdin (see below).
```
